### DISCUSSION TITLE
Migration and Type generation/mapping

### DISCUSSION THREAD

**Discussion by osaxma** [serverpod/serverpod#113] [General]
First, I’ve to say that I’m really interested and excited for this project -- mainly the ability to write in Dart from end to end and the use of Postgres. 

Jumping into the topic, I'm really having difficulty understanding the proper workflow when working with serverpod in terms of migration. I will discuss here the current state and I'll share some ideas from my humble experience.

## Current State

For instance, when starting with the example app, the example table is created within `tables.pgsql`. I believe, unless I'm missing something, that I've to manually execute the `tables.pgsql` using something like `psql` to apply the migration _(note: I don't think this step is documented)_. 

Now let's say, I wanted to add a new column/field to the example table/class. I assumed the field should be added in the `yaml` protocol file so I added it there. When running `serverpod generate`, the `tables.pgsql` file is modified to:

```sql
-- filename: tables.pgsql
CREATE TABLE example (
  "id" serial,
  "name" text NOT NULL,
  "data" integer NOT NULL,
  "newColumn" integer NOT NULL -- <~~~ this line was added
);

ALTER TABLE ONLY example
  ADD CONSTRAINT example_pkey PRIMARY KEY (id);
```

Executing this file will definitely fail because both the table and the primary key already exist. When adding a column, I would expect a new migration file containing the following statement:

```sql
ALTER TABLE example ADD COLUMN "newColumn" integer NOT NULL;
```

The file will then should be executed on its own. Since this is not the case, I assume that I've to manually write/manage the migrations as well as write to the `yaml` files under protocol in order to generate the dart classes. In other words, I have to write SQL, execute it, make sure it's executed only once, and then copy the table definition to the yaml file so Dart types can be generated. 

Anyway, I'm not sure how this could be improved while using `yaml` protocol files to define tables and types. I assume it would work if the framework can generate the corresponding sql statements based on the difference between the current state of the database, codebase and `yaml` file. I'm not sure how this would work when renaming a field, for instance, since it could be treated as deleting a field which will cause some data loss. 

I'm assuming the work on the migration part is still preliminary but I'm not sure what is the plan since I don't see migration mentioned anywhere in the readme or the roadmap. Hence, I'll take the chance to share an idea.  

## Ideas

I believe someone on reddit has already suggested using something similar to `Active Record Migration` of Ruby on Rails (RoR) which is described as follow:
> Migrations are a feature of Active Record that allows you to evolve your database schema over time. Rather than write schema modifications in pure SQL, migrations allow you to use a Ruby DSL to describe changes to your tables.
>
> source: https://guides.rubyonrails.org/active_record_migrations.html
 

I personally think their approach is neat just by looking at it and I also think Dart is capable of creating something similar. A similar approach could eliminate writing `pgsql` and `yaml` files altogether and almost everything could be written in pure Dart.

Here's an example from a project using RoR and how the migration files look like overtime: [Forem 🌱 Project](https://github.com/forem/forem/tree/main/db/migrate). The files have a datetime prefix that allow them to be sorted chronologically so migrations are applied in order. Also, I assume the database would have a table containing all the applied migrations so they are not executed more than once in a given any database (e.g. different customer or dev vs staging). 

### How that could look with Serverpod + Dart:

I'll take the example that I presented initially and I'll use some pseudo code to present the idea (at least how I imagine it since I haven't gone beyond that).

To reiterate, the goal here is to add a new column to the example table and I imagine that will take the following steps:

1. Create a new migration file 
    Using `serverpod` cli (could also be an IDE extension), run the following command:

    ```console
    user:~$ serverpod migration create add_new_column_to_example_table 
    ```

    This will create a new file with a unique chronologically sortable id (e.g. ms since epoch) under a migrations folder such as:

    ```
    ├── migrations
    │   └── 1648047284135_add_new_column_to_example_table.dart
    ```

    Inside the file, a code template is available for the user to write their migrations in dart:

    ```dart
    import 'package:serverpod/migrations.dart';

    // treat each migration as a standalone script
    void main(List<String> args) async {
        // assumptions: the arguments contain the database url (dev or staging or prod)
        final db = Database.fromArguments(args); 

        /* code written by the user start */
        await db.addColumn(tableName: 'example', columnName: 'newColumn', type: 'integer');
        /* code written by the user end */
    }
    ```

    The script above, when executed, should add the column to the table. The `Database` class should take care of that internally. 

2. Apply the migration 

    Again using the cli, the user can apply the migration:

    ```console
    user:~$ serverpod migration apply 
    ```

    Ultimately, this command should find all the unapplied migrations against a table in the database that contains the already applied migrations. Once the unapplied migrations are at hand, then the command will apply each one in the order they were written. This is important since in development, multiple migrations can be written that are later applied in bulk to staging and then production servers. 

    Running the migration should also generate/update the Dart Types from the tables on PostgreSQL which is totally doable as shown in this repository: [schema-dart](https://github.com/osaxma/schema-dart). Generating the Dart types for tables from the database itself will make the database the ultimate source of truth instead of relying on the `yaml` file. 

    Custom Dart types that are shared between backend and frontend can be created by the user using Dart directly as well instead of `yaml`. 

### Additional Example:
Here's how I imagine creating a table using the same approach: 

```dart
void main(List<String> args) async {
    final db = Database.fromArguments(args); 

    await db.createTable(
        name: 'products',
        tableBuilder: (table) {
            table.bigInt('id', primary: true, autoIncrement: true);
            table.text('username', unique: true);
            table.text('display_name');
            // example for utility/helper functions
            table.addCreatedAtColumn();  // i.e. create column + using now() as default value
            table.addUpdatedAtColumn(); // i.e. create column + update trigger 
        });
}
```

More functionality can be added such as creating foreign keys, adding indices, renaming a column, etc. 

Again, this is how I imagine it and I haven't done any actual POC. Also, I haven't thought much about how rolling back would work -- whether a combination of `up` and `down` migrations or something similar to Flyway's `Undo`. That may fundamentally change how the scripts are written. 

--- 
## Other notes:

Few things I notice while looking at the repository:
- all `serverpod` tables are in the public schema

    I noticed all the tables for serverpod are part of the public schema -- I believe they should have their separate schema(s). For instance, Supabase has a separate schema for each service -- e.g. `storage`, `auth`, `realtime` and so on. Also Hasura does a similar thing. 

- `serverpod` specific migrations 
    Similar to the discussion above, how would `serverpod` apply migrations to its own schemas/tables in the future if a change needs to be made. 

--- 

Please feel free to correct me if I made any wrong assumptions and I hope the idea will be helpful to create a developer friendly workflow for migrations and generating types. 

Cheers,
Osama

**Comment by vlidholt** [comment]
Hi @osaxma,

I just had time to have a quick look today, but will dig into this in more detail over the next few days. Also sending you a personal message on Twitter.

Best,
Viktor

**Reply by markfili** [reply]
Hi @vlidholt, any suggestions on how to best migrate manually after each change? (to use it correctly at least until the migration tool arrives)
